server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreatePartitionsAction.java

```diff
@@ -182,7 +191,30 @@ private ClusterState executeCreateIndices(ClusterState currentState, CreateParti
             String firstIndex = indicesToCreate.get(0);
             String templateName = PartitionName.templateName(firstIndex);
 
-            IndexTemplateMetadata template = currentState.metadata().templates().get(templateName);
+
+            final Metadata metadata = currentState.metadata();
+            Metadata.Builder newMetadataBuilder = Metadata.builder(metadata);
+
+            // After https://github.com/crate/crate/commit/c8edfff8fe7713b071841b9db1aea26e3f500245,
+            // we upgrade templates only on node startup in GatewayMetaState or on logical replication or on a snapshot restore.
+            // That is not enough to prevent old templates being applied on new nodes.
+
+            // Say, we are running a rolling upgrade.
+            // One node has been upgraded to VERSION.CURRENT and a master node, which is on old version < CURRENT
+            // creates partitions/new indices with old metadata/removed settings.
+            // All upgraded nodes will get incoming old template via IndicesClusterStateService
+            // and keep it forever as they already accomplished upgrade-on-startup.
+
+            // We upgrade on the fly as a trade-off to the pre-5.6.0 state
+            // where we used to upgrade template on each cluster change and submit synchronous  request via PutTemplate API to the master node to apply it.
+            // Each new node either never becomes a master and doesn't really need upgrade or eventually upgrades when it becomes a master.
+            // Since upgrade can be expensive for bulk inserts, we do it only once (per active master node).
+            // After re-election, a new master will run it again, but only once.
+            if (upgraded.compareAndSet(false, true)) {
```

Comment:
Doesn't this mean that this will only work for a single partitioned table, and if you have more than one the second will run into the original issue?